index.d.ts

```diff
@@ -0,0 +1,88 @@
+//                 Jacob Bogers <https://github.com/jacobbogers>
+//                 Naoto Yokoyama <https://github.com/builtinnya>
+//                 Ryan Cannon <https://github.com/ry7n>
+//                 HoldYourWaffle <https://github.com/HoldYourWaffle>
+
+import express from 'express'
+import node from 'events'
+
+
+
+declare global {
+  namespace Express {
+    interface Request { //inject additional properties on express.Request
+      session?: Session;
+      sessionID?: string;
+    }
+  }
+
+  export interface SessionData {
+    //seperate interface to allow application-specific declaration merging (see README)
+  }
+}
+
+
+
+interface SessionOptions {
+  secret: string | string[];
+
+  genid?(req: express.Request): string;
+  name?: string;
+  store?: Store;
+  cookie?: express.CookieOptions;
+  rolling?: boolean;
+  resave?: boolean;
+  proxy?: boolean;
+  saveUninitialized?: boolean;
+  unset?: string;
+}
+
+export default function session(options?: SessionOptions): express.RequestHandler;
+
+
+
+interface Session extends SessionData {
+  id: string;
+  cookie: SessionCookie;
+
+  regenerate(callback: (err: any) => void): Session;
+  destroy(callback: (err: any) => void): Session;
+  reload(callback: (err: any) => void): Session;
+  save(callback?: (err: any) => void): Session;
+  touch(): Session;
+}
+
+interface SessionCookie extends express.CookieOptions { //XXX actually a class
+  originalMaxAge: number;
+}
+
+
+
+export abstract class Store extends node.EventEmitter {
+  constructor(config?: any);
+
+  regenerate: (req: express.Request, callback: (err?: any) => any) => void;
+  load: (sid: string, callback: (err: any, session?: SessionData) => any) => void;
+  createSession: (req: express.Request, sess: SessionData) => void;
+
+  abstract get: (sid: string, callback: (err: any, session?: SessionData | null) => void) => void;
+  abstract set: (sid: string, session: SessionData, callback?: (err?: any) => void) => void;
+  abstract destroy: (sid: string, callback?: (err?: any) => void) => void;
+  abstract all: (callback: (err: any, obj?: { [sid: string]: SessionData; } | null) => void) => void;
+  abstract length: (callback: (err: any, length?: number | null) => void) => void;
+  abstract clear: (callback?: (err?: any) => void) => void;
+  abstract touch: (sid: string, session: SessionData, callback?: (err?: any) => void) => void;
+}
+
+export class MemoryStore extends Store {
+  get: (sid: string, callback: (err: any, session?: SessionData | null) => void) => void;
+  set: (sid: string, session: SessionData, callback?: (err?: any) => void) => void;
+  destroy: (sid: string, callback?: (err?: any) => void) => void;
+  all: (callback: (err: any, obj?: { [sid: string]: SessionData; } | null) => void) => void;
+  length: (callback: (err: any, length?: number | null) => void) => void;
+  clear: (callback?: (err?: any) => void) => void;
+  touch: (sid: string, session: SessionData, callback?: (err?: any) => void) => void;
```

Comment:
The `callback` will never return an error (impl: https://github.com/expressjs/session/blob/master/session/memory.js#L156)

Reply:
The original DT typings are even worse than I initially thought. I removed the parameter.